ingestion-beam/src/main/java/com/mozilla/telemetry/aet/DecryptAetIdentifiers.java

```diff
@@ -0,0 +1,289 @@
+      if (prevAnonIdsNode != null) {
+        payload.set("previousEcosystemUserIds", decrypt(keyStore, prevAnonIdsNode));
+      }
+    }
+
+    private void processStructuredIngestionPayload(ObjectNode json)
+        throws IOException, JoseException {
+      validator.validate(structuredAetSchema, json);
+      JsonNode anonIdNode = json.remove("ecosystem_anon_id");
+      if (anonIdNode != null) {
+        json.set("ecosystem_user_id", decrypt(keyStore, anonIdNode));
+      }
+      ArrayNode prevAnonIdsNode = (ArrayNode) json.remove("previous_ecosystem_anon_ids");
+      if (prevAnonIdsNode != null) {
+        json.set("previous_ecosystem_user_ids", decrypt(keyStore, prevAnonIdsNode));
+      }
+    }
+
+    @Override
+    public Iterable<PubsubMessage> apply(PubsubMessage message) {
+      message = PubsubConstraints.ensureNonNull(message);
+      String uri = Strings.nullToEmpty(message.getAttribute(Attribute.URI));
+      String[] uriParts = uri.split("/");
+
+      if (keyStore == null) {
+        // If configured resources aren't available, this throws UncheckedIOException;
+        // this is unretryable so we allow it to bubble up and kill the worker and eventually fail
+        // the pipeline.
+        keyStore = KeyStore.of(metadataLocation.get(), kmsEnabled.get());
+      }
+
+      if (validator == null) {
+        validator = new JsonValidator();
+        try {
+          telemetryAetSchema = JSONSchemaStore.readSchema(Resources
+              .toByteArray(Resources.getResource("account-ecosystem/telemetry.schema.json")));
+          structuredAetSchema = JSONSchemaStore.readSchema(Resources
+              .toByteArray(Resources.getResource("account-ecosystem/structured.schema.json")));
+        } catch (IOException e) {
+          // We let this problem bubble up and kill the worker.
+          throw new UncheckedIOException("Unable to load AET JSON schemas", e);
+        }
+      }
+
+      ObjectNode json;
+      try {
+        json = Json.readObjectNode(message.getPayload());
+      } catch (IOException e) {
+        throw new UnparsableAetPayloadException(e);
+      }
+
+      // Note that we must work with the raw URI because ParseUri can raise errors that would
+      // route payloads containing anon_id values to error output; this transform must be placed
+      // before ParseUri so that we can handle redacting anon_id values.
+      byte[] normalizedPayload;
+      try {
+        if (TELEMETRY_URI_PATTERN.matcher(uri).matches()) {
+          processDesktopTelemetryPayload(json);
+        } else if (false) {
+          // Placeholder condition for handling AET payloads coming from Glean-enabled applications;
+          // design is evolving in https://bugzilla.mozilla.org/show_bug.cgi?id=1634468
+        } else if (STRUCTURED_URI_PATTERN.matcher(uri).matches()) {
+          processStructuredIngestionPayload(json);
+        } else {
+          nonAetDocType.inc();
```

Comment:
We pass through the message if it doesn't match any of the expected URI patterns. This _shouldn't_ happen if the edge is routing messages correctly. Perhaps it would be better to err on the side of caution here and raise `IllegalAetPayloadException` here so the payload gets sanitized and sent to error. Or I suppose a new `IllegalAetUriException` would be more semantically appropriate.

Reply:
Sending the message to the error stream may be more actionable to diagnose the problem. Knowing what documents passed through would be useful to determine why the counter goes up. On the other hand, once the document is redacted, it becomes useless for backfill.